### ISSUE TITLE
Buildsystem refactoring and speedup by removing buildroot

### ISSUE THREAD

**Issue by cristim** [sedutil/sedutil#21]
# Problem
The builds are too slow, which is painful for local development and makes the project fail to build on  TravisCI.

I got this idea from #2, you can see the discussion there for more background/context.

Even with the newly added incremental builds introduced in #16, the initial build for local development is still really slow, and incremental builds wouldn't help on the CI build infrastructure.

# Root cause
The main culprit for the slow builds is the buildroot cross-compilation environment which we would compile from scratch, and it would be nice if we could avoid it to speed up builds significantly. 

# Proposal

On #2 I suggested to switch entirely to another embedded Linux OS based on binary packages which could be installed on top of a precompiled base, but then I did a bit of research and it turns out that TinyCore, the embedded Linux distribution used as base of the PBA, also supports additional precompiled packages just like we need to get rid of buildroot.

We could simply build TinyCore packages without using the buildroot cross-compiled toolchain. The TinyCore packages be easily compiled in a native environment running in a TinyCore Docker container image, which avoids the need to create the cross-compilation buildroot environment that makes the builds so slow.

The PBA/Rescue build would consist mainly of the following steps:
- Download a TinyCore Docker container image and install some development TinyCore packages such as gcc and make, which is what we need to build sedutil and linuxpba. We could even create a Docker image with these build dependencies preinstalled.
- Build sedutil and linuxpba into TinyCore native packages inside a  TInyCore development docker container, and leave the binary packages in the filesystem after the compilation completed.
- Download the TinyCore ISO/USB image (and cache it locally) and remaster it by introducing the newly created sedutil packages. The only difference between the Rescue and PBA images is which packages we install and how we configure the init scripts (the PBA image needs some custom init configuration which we already have, in order to run the password prompt at boot). We could create both images really fast based on these prebuilt binaries.

All these steps should be pretty fast since we remove the painfully slow buildroot build process which is actually building a full cross-compilation environment, and replace it with a TInyCore native build environment running in a Docker container.

# Benefits
- The speed gain would make sedutil more developer-friendly
- It would allow us to run builds from CI. 
- This approach could be reused later on if we decide to rewrite sedutil/linuxpba into a new tool  leveraging the latest Linux Kernel 4.12+ OPAL support and nvme-cli.

**Comment by cristim** [comment]
Started to work on this in a branch, you can see progress at https://github.com/cristim/sedutil/tree/buildroot-removal

So far I managed to get the software built inside the TinyCore Docker image, avoiding the buildroot entirely.

Nest steps:
- create TinyCore packages out of the binaries and leave them in the filesystem, as per http://wiki.tinycorelinux.net/wiki:creating_extensions
- remaster the TinyCore vanilla image to include these packages, obtaining the PBA and Rescue images, as per http://wiki.tinycorelinux.net/wiki:remastering_with_qremaster

**Comment by davidepic91** [comment]
In my opinion, at the end of this process, it is useful to procuce some documentation/tool in order to make easier the process of adding other libraries to the TinyCore Docker image.

**Comment by cristim** [comment]
I'm not sure how much we want it to be customized by users, but since it's going to be all done leveraging Tinycore, we can just point to their remastering instructions

**Comment by OliverO2** [comment]
In my opinion this idea is heading in the right direction.

Some remarks:
* The current build system is way too convoluted. It would be ideal to have just one boot image for everything (PBA and rescue system).
* Not having to build entire kernels would be a big plus. It's a waste of time to build something individually that's readily available elsewhere.
* The current Tinycore-based rescue system boots via BIOS only (no UEFI). Does Tinycore actually support EFI boot?
* When I tried to figure out which keymaps to use, I found the Tinycore documentation lacking the necessary depth.
* I don't think a PBA or rescue system needs customization beyond basic stuff (keyboard support. for example). For everything else users can use secondary boot images or emergency systems as they like after drives have been unlocked.

**Comment by cristim** [comment]
The only difference between the images would be the init configuration, the rest of the image content can be identical. But is is convenient for the Rescue image to also contain a copy of the PBA image for easy configuration and to have maybe a more user-friendly way to set up the encryption.

As far as I know the current UEFI images are also based on TinyCore. I haven't yet looked into the details, but there may be some more remastering needed for the UEFI boot.

The keymap support could be all baked in and configured in multiple image flavors, much like for example Firefox gives installation packages for different languages. Other ways may be to somehow write it to the CMOS settings (if doable at all) or to somehow alter a unique prebuilt PBA images right before writing it into the OPAL PBA in order to configure the keymap.

**Comment by OliverO2** [comment]
Including a PBA copy in the rescue image sounds reasonable. This could be done by remastering a universal boot image to include a copy of itself. The remastered flavor would then become the rescue image.

Currently the PBA images are not based on TinyCore but on buildroot and syslinux. Only the rescue image is based on TinyCore. And it seems that TinyCore [supports UEFI boot since version 6.2](http://forum.tinycorelinux.net/index.php?topic=18841.0).

Currently, all keymaps supported by TinyCore are already included in the PBA and rescue images. The only configuration necessary is to select the right one and provide it to `loadkmap`. This could be done by remastering a universal PBA image it with a user-selectable keymap before installation. The rescue image could also ask for the required keyboard layout at boot time. For PBA images, I'd consider it too annoying to ask the user on each boot so it's better to have it baked in at installation time. I'd rather not fiddle with CMOS.

To sum it up, we could just ship one universal boot image based on TinyCore along with an installation script to generate the required flavors (PBA and rescue images for various keyboard layouts). To make commissioning a drive for encryption easier and safer, we could extend that installation script to perform all the necessary steps and checks (selecting the drive, locking it, uploading the PBA).

**Comment by cristim** [comment]
Thanks @OliverO2, this nailed it all and sounds great!

I wasn't aware the PBA isn't based on TinyCore, but it will be great to have it rebased and enhanced as per your comment.

**Comment by OliverO2** [comment]
Hi @cristim, I'd like to update you on recent developments in this area.

I have integrated PBA creation and easy-to-use SED setup into Relax-and-Recover (ReaR), the leading Open Source bare metal disaster recovery and system migration solution. [It has been merged](https://github.com/rear/rear/pull/1659) into the current master branch of https://github.com/rear/rear.

What you'll get is a PBA created directly from your local Linux flavor. It comes with local keyboard support, retries without reboot after mis-typing an unlock password, and, if the Plymouth boot splash is installed, a nice user interface. What you'll also get is an easy-to-use tool for SED setup and administration. And, last not least, ReaR's main purpose: a complete rescue system for disaster recovery.

If you want to try it out,

- make sure that [DTA's `sedutil-cli` (version 1.15.1)](https://github.com/Drive-Trust-Alliance/exec/blob/master/sedutil_LINUX.tgz?raw=true) is available in root’s search path,
- clone or download the [current master branch of rear/rear](https://github.com/rear/rear),
- change your working directory to the rear base directory,
- create a file `etc/rear/site.conf` with the following line:
```
OUTPUT=RAWDISK
```
- run `sudo usr/sbin/rear -v mkopalpba`,
- find the PBA image in `var/lib/rear/TCG-Opal-PBA/$HOSTNAME/TCG-Opal-PBA-$HOSTNAME.raw`,
- `dd` the PBA image onto your USB stick for testing.

[Chapter 13 of the User Guide](https://github.com/rear/rear/blob/master/doc/user-guide/13-tcg-opal-support.adoc) explains how to use the PBA and administer SEDs.

**Comment by cristim** [comment]
Thanks, that's awesome!
